SU2_CFD/src/fluid/CSU2TCLib.cpp

```diff
@@ -2004,39 +2003,78 @@ vector<su2double>& CSU2TCLib::ComputeTemperatures(vector<su2double>& val_rhos, s
   T = (rhoE - rhoEve - rhoE_f + rhoE_ref - rhoEvel) / rhoCvtr;
 
   /*--- Set temperature clipping values ---*/
-  su2double Tmin  = 50.0; su2double Tmax = 8E4;
-  su2double Tve_o = 50.0; su2double Tve2 = 8E4;
+  su2double Tmin   = 50.0; su2double Tmax   = 8E4;
+  su2double Tvemin = 50.0; su2double Tvemax = 8E4;
+  su2double Tve_o  = 50.0; su2double Tve2  = 8E4;
 
   /* Determine if the temperature lies within the acceptable range */
   if      (T < Tmin) T = Tmin;
   else if (T > Tmax) T = Tmax;
 
   /*--- Set vibrational temperature algorithm parameters ---*/
+  su2double NRtol         = 1.0E-6;    // Tolerance for the Newton-Raphson method
   su2double Btol          = 1.0E-6;    // Tolerance for the Bisection method
   unsigned short maxBIter = 50;        // Maximum Bisection method iterations
+  unsigned short maxNIter = 50;        // Maximum Newton-Raphson iterations
+  su2double scale         = 0.9;       // Scaling factor for Newton-Raphson step
 
+  /*--- Execute a Newton-Raphson root-finding method for Tve ---*/
   //Initialize solution
   Tve = T;
 
-  // Execute the root-finding method
   bool Bconvg = false;
-  su2double rhoEve_t = 0.0;
-
-  for (unsigned short iIter = 0; iIter < maxBIter; iIter++) {
-    Tve      = (Tve_o+Tve2)/2.0;
-    val_eves = ComputeSpeciesEve(Tve);
-    rhoEve_t = 0.0;
-    for (iSpecies = 0; iSpecies < nSpecies; iSpecies++) rhoEve_t += rhos[iSpecies] * val_eves[iSpecies];
-    if (fabs(rhoEve_t - rhoEve) < Btol) {
-      Bconvg = true;
-      break;
+  bool NRconvg = false;
+  su2double rhoEve_t = 0.0, rhoCvve = 0.0;
+  vector<su2double> val_eves;
+  vector<su2double> val_cvves;
+
+  /*--- Newton-Raphson Method --*/
+  for (unsigned short iIter = 0; iIter < maxNIter; iIter++) {
+    rhoEve_t = rhoCvve = 0.0;
+    val_eves  = ComputeSpeciesEve(Tve);
```

Comment:
If these functions return references you should use const auto& "X" = "f" to avoid copies in this loop.